### PR TITLE
perf(store): optimize block cache reclaiming

### LINKED ISSUE #819: Polish the block cache reclaim policy
There is a known issue with the current block cache implementation that causes the reclaim process to take tens of milliseconds. The primary reasons for this are as follows:

* Inefficiency in sorting cached entries: When the total cache size is large or when there are numerous small cached entries, the current implementation sorts them to find the best candidate for reclamation, resulting in low efficiency.

* Reclaim policy based on watermark: The current reclaim policy is based on a watermark, where each reclaim attempt tries to reclaim the cache from a high to a low watermark. This leads to the simultaneous reclamation of a large amount of memory at once.

Possible solutions for optimization:

* Reclaim based on a specific value: Instead of relying on a watermark, consider reclaiming cache based on a specific value. Fortunately, the uring mod already has knowledge of the exact memory usage when performing batched IOs, which can be leveraged for more efficient reclamation.

* Implement an LRU data structure for each block cache segment: Adding an LRU (Least Recently Used) data structure to each block cache of the segment allows for dynamic adjustment of the reclaim priority of each cached entry. This way, the most suitable entries can be chosen for reclamation, improving efficiency.

* Simplify the reclaim function of the WAL: The reclaim function of the WAL (Write-Ahead Log) can be optimized by eliminating the need to sort all cached entries. Instead, implement on-demand reclamation through a simple merge sort, which can significantly reduce the overhead and improve performance.

By incorporating these methods, the block cache implementation can be enhanced to address the current issue and achieve better performance.

### PR DESCRIPTION
**Pull request by ShadowySpirits**
fix #819

### CODE REVIEW

File: components/store/src/io/wal.rs [hunk 1]
```diff
@@ -433,18 +429,19 @@ impl Wal {
...
 
-        // Calculate the bytes to reclaim, based on the configured `high_water_mark` and `low_water_mark`.
-        let high_percent = Percentage::from(self.wal_cache.high_water_mark);
-        let low_percent = Percentage::from(self.wal_cache.low_water_mark);
+        // Calculate the bytes to reclaim, based on the configured `high_watermark`.
+        let high_percent = Percentage::from(self.wal_cache.high_watermark);
         let mut to_reclaim = 0;
         if cache_size > high_percent.apply_to(self.wal_cache.max_cache_size) {
-            to_reclaim = cache_size - low_percent.apply_to(self.wal_cache.max_cache_size);
+            to_reclaim = cache_size - high_percent.apply_to(self.wal_cache.max_cache_size);
```
Comment (ShadowySpirits) [inline components/store/src/io/wal.rs:444]: @lizhanhui @daniel-y I prefer to reclaim more data to reduce reclaiming operation. Do you have any suggestions?

File: components/store/src/io/block_cache.rs [hunk 2]
```diff
@@ -213,6 +223,26 @@ impl BlockCache {
...
+                        }
+
+                        // And then compare by score and wal offset.
+                        let score_order = entry_a.score().cmp(&entry_b.score());
+                        if score_order == Ordering::Equal {
+                            return entry_a.wal_offset().cmp(&entry_b.wal_offset());
+                        }
+                        score_order
+                    },
+                )
```
[the remark this answers is not in the document]
Comment (ShadowySpirits) [reply components/store/src/io/block_cache.rs:244]: We should ensure entry with the same wal offset won't add in score list duplicated. The compare function is transitive under this condition.

File: components/store/src/io/block_cache.rs [hunk 3]
```diff
@@ -213,6 +223,26 @@ impl BlockCache {
         Self {
             wal_offset: offset,
             entries: BTreeMap::new(),
+            score_list: unsafe {
```
Comment (lizhanhui) [inline components/store/src/io/block_cache.rs:226]: Why unsafe is required?

File: components/store/src/io/block_cache.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Chillax-0v0) [reply components/store/src/io/block_cache.rs:244]: If `entry_a.wal_offset() == entry_b.wal_offset()` will never happen, I think L233-235 can be removed, or just panic here.

File: components/store/src/io/block_cache.rs (hunk 3 above)
Comment (ShadowySpirits) [reply components/store/src/io/block_cache.rs:226]: OrderedSkipList::with_comp is an unsafe function.

**Comment by lizhanhui** [comment]
Let's merge this issue first and make changes a secondary PR.

File: components/store/src/io/block_cache.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ShadowySpirits) [reply components/store/src/io/block_cache.rs:244]: Ordering::Equal is required by the OrderedSkipList::remove function.

**Pull request merged**